src/__tests__/metrics.test.ts

```diff
@@ -100,73 +103,59 @@ describe('Metrics Query', () => {
-      );
-      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
+
+    beforeAll(async () => {
+      backendSrvMock.datasourceRequest = jest
+        .fn()
+        .mockImplementation(() => Promise.resolve(assetsResponse));
+      result = await ds.metricFindQuery(variableQuery);
     });
-  });
 
-  describe('Given an incorrect metrics query', () => {
-    const variableQuery: VariableQueryData = {
-      query: 'asset{name=~asset.*}',
-      filter: '',
-    };
-    it('should throw a parse error', () => {
-      expect(ds.metricFindQuery(variableQuery)).rejects.toThrowErrorMatchingInlineSnapshot(
-        `"ERROR: Unable to parse 'name=~asset.*'. Only strict equality (=) is allowed."`
-      );
-      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
+    it('should generate the correct request', () => {
+      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
+      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
     });
-  });
 
-  describe('Given an incorrect metrics query', () => {
-    const variableQuery: VariableQueryData = {
-      query: 'asset{name="asset}',
-      filter: '',
-    };
-    it('should throw a parse error', () => {
-      expect(ds.metricFindQuery(variableQuery)).rejects.toThrowErrorMatchingInlineSnapshot(
-        `"ERROR: Could not find closing ' \\" ' while parsing 'name=\\"asset'."`
-      );
-      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
+    it('should return the correct assets', () => {
+      expect(result.length).toEqual(0);
     });
   });
 
-  describe('Given an incorrect filter query', () => {
+  describe('Given an incomplete metrics query', () => {
     const variableQuery: VariableQueryData = {
-      query: 'asset{name=foo}',
-      filter: 'filter{',
+      query: 'asset',
     };
-    it('should throw a filter error', () => {
-      expect(ds.metricFindQuery(variableQuery)).rejects.toThrowErrorMatchingInlineSnapshot(
-        `"ERROR: Unable to parse expression filter{"`
-      );
+    it('should return an empty array', async () => {
+      const result = await ds.metricFindQuery(variableQuery);
+      expect(result).toEqual([]);
       expect(backendSrvMock.datasourceRequest).not.toBeCalled();
     });
   });
 
   describe('Given an incorrect filter query', () => {
-    const variableQuery: VariableQueryData = {
-      query: 'asset{name=foo}',
-      filter: 'filter{foo}',
-    };
-    it('should throw a filter error', () => {
-      expect(ds.metricFindQuery(variableQuery)).rejects.toThrowErrorMatchingInlineSnapshot(
-        `"ERROR: Could not parse: 'foo'. Missing a comparator (=,!=,=~,!~)."`
-      );
+    it('should return an empty array if filter type is wrong', async () => {
+      const variableQuery: VariableQueryData = {
+        query: "assets{name~='foo'}",
+      };
+      const result = await ds.metricFindQuery(variableQuery);
+      expect(result).toEqual([]);
```

Comment:
`metricFindQuery` is using while you creating a new variable. If you'll just throw error it will not change previously found result in preview section.

For now, when you're adding a variable all errors are handled by variable controller and renders message at the bottom of query string.

Reply:
did you try to look at it?
what I am curious about is how hard is that to throw an error, catch it properly and replace value with [] in a ctrl?

Reply:
or it doesn't make it better?

Reply:
the main thing that `metricFindQuery` returns a value that will be displayed in a preview section of variable adding component. I can control what will be displayed in the preview section only by returning some value from `metricFindQuery`

Reply:
result of conversation: not possible to fix rn, because of some grafana behaviour